File: CSC310-Cpp/_solutions/Module15-Computational-Geometry/debug01/segment_intersection.cpp
// segment_intersection.cpp
// Decides whether two line segments intersect (CLRS 33.1) using cross
// products only, with no division and no floating point. Two segments
// intersect if they share at least one point, including touching at an
// endpoint or overlapping along a line.
#include <algorithm>
#include <iostream>
#include <string>

struct Point {
    long x, y;
};

struct Segment {
    Point p, q;
};

// Sign of the cross product (b - a) x (c - a): positive when a->b->c turns
// left (counter-clockwise), negative when it turns right, zero when collinear.
long direction(Point a, Point b, Point c) {
    return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
}

// Given that c is collinear with a and b, is c on the segment a-b?
bool onSegment(Point a, Point b, Point c) {
    return std::min(a.x, b.x) <= c.x && c.x <= std::max(a.x, b.x) &&
           std::min(a.y, b.y) <= c.y && c.y <= std::max(a.y, b.y);
}

bool intersects(const Segment& s1, const Segment& s2) {
    long d1 = direction(s2.p, s2.q, s1.p);
    long d2 = direction(s2.p, s2.q, s1.q);
    long d3 = direction(s1.p, s1.q, s2.p);
    long d4 = direction(s1.p, s1.q, s2.q);
    bool s1Straddles = (d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0);
    bool s2Straddles = (d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0);
    if (s1Straddles && s2Straddles) {
        return true;
    }
    // Boundary cases: an endpoint of one segment lies on the line of the other.
    if (d1 == 0 && onSegment(s2.p, s2.q, s1.p)) return true;
    if (d2 == 0 && onSegment(s2.p, s2.q, s1.q)) return true;
    if (d3 == 0 && onSegment(s1.p, s1.q, s2.p)) return true;
    if (d4 == 0 && onSegment(s1.p, s1.q, s2.q)) return true;
    return false;
}

void report(const std::string& label, Segment a, Segment b) {
    std::cout << label << ": (" << a.p.x << "," << a.p.y << ")-(" << a.q.x << "," << a.q.y
              << ") and (" << b.p.x << "," << b.p.y << ")-(" << b.q.x << "," << b.q.y << ") -> "
              << (intersects(a, b) ? "intersect" : "disjoint") << '\n';
}

int main() {
    report("crossing        ", {{0, 0}, {4, 4}}, {{0, 4}, {4, 0}});
    report("T junction      ", {{0, 0}, {4, 0}}, {{2, 0}, {2, 3}});
    report("shared endpoint ", {{0, 0}, {3, 3}}, {{3, 3}, {6, 0}});
    report("collinear overlap", {{0, 0}, {4, 0}}, {{2, 0}, {6, 0}});
    report("collinear apart ", {{0, 0}, {2, 0}}, {{3, 0}, {5, 0}});
    report("parallel        ", {{0, 0}, {4, 0}}, {{0, 1}, {4, 1}});
    report("near miss       ", {{0, 0}, {4, 4}}, {{5, 0}, {6, 3}});
    return 0;
}

File: CSC310-Python/Module15-Computational-Geometry/debug01/segment_intersection.py
"""Line-segment intersection test (CLRS Section 33.1).

segments_intersect(p1, p2, p3, p4) decides whether segment p1p2 and
segment p3p4 share at least one point, using only cross products: no
division and no floating point.  Two segments intersect if they cross
properly, if an endpoint of one lies on the other, if they share an
endpoint, or if they overlap along the same line.  The main block tests
eight configurations.
"""


def direction(pi, pj, pk):
    """Cross product (pk - pi) x (pj - pi): the turn direction at pi."""
    return (pk[0] - pi[0]) * (pj[1] - pi[1]) - (pj[0] - pi[0]) * (pk[1] - pi[1])


def on_segment(pi, pj, pk):
    """Given that pk is collinear with pi and pj, is pk between them?"""
    return (min(pi[0], pj[0]) < pk[0] < max(pi[0], pj[0]) and
            min(pi[1], pj[1]) < pk[1] < max(pi[1], pj[1]))


def segments_intersect(p1, p2, p3, p4):
    d1 = direction(p3, p4, p1)
    d2 = direction(p3, p4, p2)
    d3 = direction(p1, p2, p3)
    d4 = direction(p1, p2, p4)
    if ((d1 > 0 and d2 < 0) or (d1 < 0 and d2 > 0)) and \
       ((d3 > 0 and d4 < 0) or (d3 < 0 and d4 > 0)):
        return True
    if d1 == 0 and on_segment(p3, p4, p1):
        return True
    if d2 == 0 and on_segment(p3, p4, p2):
        return True
    if d3 == 0 and on_segment(p1, p2, p3):
        return True
    if d4 == 0 and on_segment(p1, p2, p4):
        return True
    return False


def main():
    cases = [
        ("crossing", (0, 0), (4, 4), (0, 4), (4, 0)),
        ("T junction", (0, 0), (4, 0), (2, 0), (2, 3)),
        ("shared endpoint", (0, 0), (3, 3), (3, 3), (6, 0)),
        ("collinear overlap", (0, 0), (4, 0), (2, 0), (6, 0)),
        ("diagonal overlap", (0, 0), (3, 3), (2, 2), (5, 5)),
        ("collinear apart", (0, 0), (2, 0), (3, 0), (5, 0)),
        ("parallel", (0, 0), (4, 0), (0, 1), (4, 1)),
        ("near miss", (0, 0), (4, 4), (5, 0), (6, 3)),
    ]
    for name, p1, p2, p3, p4 in cases:
        verdict = "intersect" if segments_intersect(p1, p2, p3, p4) else "disjoint"
        print(f"{name:<18}: {p1}-{p2} and {p3}-{p4} -> {verdict}")


if __name__ == "__main__":
    main()
